core/src/main/java/com/datastax/oss/driver/api/core/AsyncPagingIterable.java

```diff
@@ -0,0 +1,102 @@
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.Nullable;
+import java.util.Iterator;
+import java.util.concurrent.CompletionStage;
+import java.util.function.Function;
+
+/**
+ * An iterable of elements which are fetched asynchronously by the driver, possibly in multiple
+ * requests.
+ */
+public interface AsyncPagingIterable<ElementT> {
+
+  /** Returns {@linkplain ExecutionInfo information about the execution} of this page of results. */
+  @NonNull
+  ExecutionInfo getExecutionInfo();
+
+  /** How many rows are left before the current page is exhausted. */
+  int remaining();
+
+  /**
+   * The elements in the current page. To keep iterating beyond that, use {@link #hasMorePages()}
+   * and {@link #fetchNextPage()}.
+   *
+   * <p>Note that this method always returns the same object, and that that object can only be
+   * iterated once: elements are "consumed" as they are read.
+   */
+  @NonNull
+  Iterable<ElementT> currentPage();
+
+  /**
+   * Returns the next element, or {@code null} if the results are exhausted.
+   *
+   * <p>This is convenient for queries that are known to return exactly one element, for example
+   * count queries.
+   */
+  @Nullable
+  default ElementT one() {
+    Iterator<ElementT> iterator = currentPage().iterator();
+    return iterator.hasNext() ? iterator.next() : null;
+  }
+
+  /**
+   * Whether there are more pages of results. If so, call {@link #fetchNextPage()} to fetch the next
+   * one asynchronously.
+   */
+  boolean hasMorePages();
+
+  /**
+   * Fetch the next page of results asynchronously.
+   *
+   * @throws IllegalStateException if there are no more pages. Use {@link #hasMorePages()} to check
+   *     if you can call this method.
+   */
+  @NonNull
+  CompletionStage<? extends AsyncPagingIterable<ElementT>> fetchNextPage()
+      throws IllegalStateException;
+
+  /**
+   * If the query that produced this result was a conditional update, indicate whether it was
+   * successfully applied.
+   *
+   * <p>In theory this method should not be exposed on {@link AsyncPagingIterable} since it is
+   * specific to CQL; in practice, most {@code AsyncPagingIterable} instances will be created by
+   * wrapping an {@link AsyncResultSet}, so it is convenient to make this information available
+   * without having to unwrap.
+   *
+   * @see AsyncResultSet#wasApplied()
+   */
+  boolean wasApplied();
+
+  /**
+   * Creates a new instance by transforming each element of this iterable with the provided
+   * function.
+   *
+   * <p>Note that both instances share the same underlying data: consuming elements from the
+   * transformed iterable will also consume them from this object, and vice-versa.
+   */
+  default <TargetT> AsyncPagingIterable<TargetT> map(
+      Function<? super ElementT, ? extends TargetT> elementMapper) {
+    return new AsyncPagingIterableWrapper<>(this, elementMapper);
+  }
```

Comment:
I really like to have this here. It's the most natural thing that users might want to do with a result set, so it makes sense to have a shortcut for it. Also it will be super convenient in the object mapper.

As I said in JIRA, I think we should stay minimal and not try to implement other functional operators. They don't necessarily work well with the concept of page boundaries.

Reply:
I'm not sure actually. First, it's already trivial to obtain a Java 8 stream from `currentPage()` and call `map()` on it; the same is valid for reactive streams, so on-the-fly mapping use cases are imo already well covered. 
For users wanting to transform a result set A into a result set B in one single call, other solutions exist: a) write a custom wrapper (similar to `AsyncPagingIterableWrapper `) or b) register a custom request processor. So in the end I'm not sure that this method buys us much.

Reply:
> it's already trivial to obtain a Java 8 stream from `currentPage()` and call `map()` on it

If you work with async result sets, you still have to write the logic to rewrap every subsequent page, which incurs a bit of boilerplate.
If you use synchronous ones, you can indeed map the iterator, but you lose the paging methods (`isFullyFetched`, `getAvailableWithoutFetching`).

What's nice with `map()` is that you get all of that for free.

> write a custom wrapper (similar to `AsyncPagingIterableWrapper`)

Yes, that's pretty easy to write. But we're going to need it for the mapper anyway, so why not make it available?

> register a custom request processor

That's the nuclear option 😉

Reply:
> but you lose the paging methods (isFullyFetched, getAvailableWithoutFetching)

Hmm OK this finally convinced me of the usefulness of the `map` method.